src/bit-fields-macros/src/builder.rs

```diff
@@ -0,0 +1,816 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+use std::fmt::Write;
+
+use proc_macro2::{Ident, TokenStream};
+use quote::quote;
+
+use crate::utils::{DataTypeToken, MultiLineString};
+use crate::{Field, Flag, Member};
+
+/// Builder for bit fields.
+#[allow(clippy::module_name_repetitions)]
+#[derive(Debug)]
+pub struct BitFieldBuilder {
+    /// String used to define `From<HashSet<String>>`.
+    #[cfg(feature = "flag_set")]
+    flag_matching_from_hashset: TokenStream,
+    /// String used to define `From<HashSet<String>>`.
+    #[cfg(feature = "flag_set")]
+    flag_setting_hashset: TokenStream,
+    /// String used to define `From<HashMap<String,$data_type>>`.
+    #[cfg(feature = "field_map")]
+    field_matching_from_hashmap: TokenStream,
+    /// String used to define `From<HashMap<String,$data_type>>`.
+    #[cfg(feature = "field_map")]
+    field_setting_hashmap: TokenStream,
+    /// String used to define `superset()`.
+    #[cfg(feature = "set_theory")]
+    fields_superset_fn: TokenStream,
+    /// String used to define `subset()`.
+    #[cfg(feature = "set_theory")]
+    fields_subset_fn: TokenStream,
+    /// String used to define `disjoint()`.
+    #[cfg(feature = "set_theory")]
+    fields_disjoint_fn: TokenStream,
+    /// String used to define `intersection()`.
+    #[cfg(feature = "set_theory")]
+    fields_intersection_fn: TokenStream,
+    /// String used to define `union()`.
+    #[cfg(feature = "set_theory")]
+    fields_union_fn: TokenStream,
+    /// String used to define the table used in the rustdoc for the bit field.
+    struct_doc_table_layout: String,
+    /// Accessor methods to members.
+    struct_accessors: TokenStream,
+    /// String used to form the display the bit field, the lines represent:
+    /// 1. Top border
+    /// 2. Bit numbers
+    /// 3. Border
+    /// 4. Field idents
+    /// 5. Border
+    /// 6. Field values
+    /// 7. Bottom border
+    /// Fmt values (since write doesn't work with in place ones)
+    display_string: MultiLineString,
+    /// String used to pass arguments for `std::fmt::Display` implementation.
+    display_fmt_string: TokenStream,
+    /// Struct data type (e.g. `u8`)
+    data_type: DataTypeToken,
+    /// Struct identifier
+    struct_name: Ident,
+    /// `BitField::equal`
+    equal_defined_bits: TokenStream,
```

Comment:
This currently results in code as follows:

```rs
fn equal(&self, other: &Self) -> bool {
            true && self.RANGE1() == other.RANGE1() && self.SSE() == other.SSE()
                && self.SSE1() == other.SSE1() && self.RANGE2() == other.RANGE2()
                && self.SSE2() == other.SSE2() && self.SSE3() == other.SSE3()
                && self.RANGE3() == other.RANGE3() && self.SSE4() == other.SSE4()
        }
```

If you instead kept an associated constant `MASK` that is 1 for every bit mapped to a Bit or BitRange, and 0 for all other, you could simple do `self.0 & MASK == other.0 & MASK`. Similarly, a lot of the other set theoretic functions simplify drastically. 

Also, why isn't this the `PartialEq` implementation? Is there a usecase where the value of unmapped bits could be important for determining equality?

Reply:
> If you instead kept an associated constant MASK that is 1 for every bit mapped to a Bit or BitRange, and 0 for all other, you could simple do self.0 & MASK == other.0 & MASK. Similarly, a lot of the other set theoretic functions simplify drastically.

Just did this, it has simplified a lot.

> Also, why isn't this the PartialEq implementation? Is there a usecase where the value of unmapped bits could be important for determining equality?

The current use case for checking the value of un-mapped bits is the typical usage. It is more efficient to compare the underlying integers than to apply a mask then compare them. Typically you won't be manipulating unmapped bits so it should be sufficient to compare full equality, however the case exists for avoiding unmapped bits and this is why this function exists.

Reply:
I see, thanks for clarifying! Maybe add a comment to `.equal` (or even the BitField struct) highlighting the difference to `PartialEq`?

Although, if the usecase for comparing even the unmapped bits is the less common one, we should make `PartialEq` compare only the mapped bits, and provide a special method for also comparing the unmapped ones? Or point the user to `field1.0 == field2.0` for the latter?

Reply:
See https://github.com/firecracker-microvm/firecracker/pull/3105#discussion_r1027891071

Reply:
Yep, looks good (sorry, I misunderstood your comment when reading it the first time. I thought you meant that the normal usecase is to _not_ check the value of unmapped bits. My bad)